File: Tutorial-1/comp-obj-tut-1.jl
using DifferentialEquations
using Printf
using Plots
using LaTeXStrings
gr()
# define the set of odes
function stellar_structure!(du, u, params, m)
    a, b, n, s, nu, lambda, A = params
    p, x, t, l = u
    du[1] = - (m / x^4)
    du[2] = (t^b / (x^2 * p^a))
    du[3] = - (p^(a * n) * l) / (x^4 * t^(3 + s + b * n))
    du[4] = A * p^(a * lambda) * t^(nu - (b * lambda))
end

# We define the condition to find the event where p becomes less than some tolerace set by user.
condition(u, m, integrator) =  u[1] - 1e-4
affect!(integrator) = terminate!(integrator)
cb = ContinuousCallback(condition, affect!)

# Initial conditions, at m = 0, as per eq 1.2
p0 = 1.0
x0 = 1e-4
t0 = 1.0
l0 = 0.0

u0 = [p0, x0, t0, l0]

# parameters
a = 1.0
b = 1.0
n = 0.0
s = 0.0
nu = 4.0
lambda = 1.0
A = 0.535

params = [a, b, n, s, nu, lambda, A]
mspan = (0, 20.0)
problem1 = ODEProblem(stellar_structure!, u0, mspan, params)
solution1 = solve(problem1, alg_hints=[:stiff], callback=cb)
println(solution1.t[end], "\t", solution1.u[end][1])
plot(solution1, vars=(0, 1), lw = 2, xlabel = L"m", label = L"p")
plot!(solution1, vars=(0, 2), lw = 2, xlabel = L"m", label = L"x")
plot!(solution1, vars=(0, 3), lw = 2, xlabel = L"m", label = L"t")
plot!(solution1, vars=(0, 4), lw = 2, xlabel = L"m", label = L"l")
plot!(title = L"A=%$A", titlefontsize = 10, xlims = [0, 12], ylims = [0, 5])
savefig("p_vs_m_A_$A.pdf")

File: Tutorial-1/comp-obj-tut-1.py
import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt
from figStyle import plot_for


def stellar_structure(m, y, params):
    a, b, n, s, nu, lam, A = params
    p, x, t, ell = y

    dpdm = - (m / x**4)
    dxdm = (t**b / (x**2 * p**a))
    dtdm = - (p**(a*n) * ell) / (x**4 * t**(3 + s + b * n))
    dldm = A * p**(a * lam) * t**(nu - (b * lam))

    return [dpdm, dxdm, dtdm, dldm]


# In the first part of the problem we set A to 1.
# The initial values at the center of the start could be read off from eq 1.2
a = 1.0
b = 1.0
n = 0.0
s = 0.0
lam = 1.0
nu = 4.0
A = 0.534
params = [a, b, n, s, nu, lam, A]
mspan = [0, 100.0]
m_eval = np.arange(mspan[0], mspan[-1], 0.001)
y0 = [1.0, 1e-7, 1.0, 0.0]


# We want to find where p(m) goes to zero. So we define an event function
def hit_surface(m, y, params):
    return y[0] - 1e-3


hit_surface.terminal = True
hit_surface.direction = -1

sol1 = solve_ivp(stellar_structure, mspan, y0, t_eval=m_eval, args=(params,),
                 events=hit_surface)
mstar = sol1.t_events
# print(mstar)
print(sol1.t[-1], sol1.y[0][-1], sol1.y[3][-1])
# Let's also plot p(m) vs m to see where it becomes zero.
fig, ax = plt.subplots(nrows=1, ncols=1)
variables = ["$p$", "$x$", "$t$", "$l$"]
styles = ["solid", "dashed", "dotted", "dashdot"]
for idx, (style, label) in enumerate(zip(styles, variables)):
    ax.plot(sol1.t, sol1.y[idx], ls=style, label=rf"{label}")

ax.set_xlim(0, round(sol1.t[-1]) + 1)
ax.set_ylim(0, 2)
ax.set_xlabel(r"$m$")
ax.grid()
ax.legend()
fig.savefig(f"./p_vs_m_A_{A}.pdf")
